crates/matrix-sdk/src/client/builder.rs

```diff
@@ -522,10 +482,64 @@ impl ClientBuilder {
     }
 }
 
+/// Discovers a homeserver from a server name or a URL.
+///
+/// Tries well-known discovery and checking if the URL points to a homeserver.
+async fn discover_homeserver_from_server_name_or_url(
+    server_name_or_url: String,
+    http_client: &HttpClient,
+) -> Result<(String, Option<discover_homeserver::Response>), ClientBuildError> {
+    // Attempt discovery as a server name first.
+    let sanitize_result = sanitize_server_name(&server_name_or_url);
+    if let Ok(server_name) = sanitize_result.as_ref() {
+        let protocol = if server_name_or_url.starts_with("http://") {
+            UrlScheme::Http
+        } else {
+            UrlScheme::Https
+        };
+
+        let discovery_error =
+            match discover_homeserver(server_name.clone(), protocol, http_client).await {
+                Ok(well_known) => {
+                    return Ok((well_known.homeserver.base_url.clone(), Some(well_known)));
+                }
+                Err(e) => {
+                    debug!(error = %e, "Well-known discovery failed.");
+                    e
+                }
+            };
+
+        // Check if the server name points to a homeserver.
+        let homeserver = match protocol {
+            UrlScheme::Http => format!("http://{server_name}"),
+            UrlScheme::Https => format!("https://{server_name}"),
+        };
+
+        if let Ok(homeserver_url) = Url::parse(&homeserver) {
+            if check_is_homeserver(&homeserver_url, http_client).await {
+                return Ok((homeserver_url.to_string(), None));
+            }
+        }
+
+        return Err(discovery_error);
```

Comment:
There's a bit of code duplication here, and I think we could get rid of it:

- make `server_name_or_url` mut and replace it with the right-hand side expression of the `homeserver` assignment above
- stash the discovery error in an `Option`
- instead of returning early line 524, fall through to the next branch
- at the final error line 536, return `Err(discovery_error.or(ClientBuildError::InvalidServerName))`

How does that sound? Maybe the duplication is not too bad here, but it took me a while to figure this was doing something new...

Reply:
I was torn because of the duplication too, your proposition seems good.

Reply:
Done